File: frontend/lib/api/assets.ts
import { invoke } from '@tauri-apps/api/core'
import type { Asset, NewAssetInput, AssetSummary, ApiResponse } from '@/lib/types'

export async function getAssets(): Promise<ApiResponse<Asset[]>> {
  try {
    const assets = await invoke('list_assets') as Asset[]
    return { success: true, data: assets }
  } catch (error) {
    console.error('Failed to load assets:', error)
    return { success: false, error: 'Failed to load assets' }
  }
}

export async function getAssetSummary(): Promise<ApiResponse<AssetSummary>> {
  try {
    const summary = await invoke('get_asset_summary') as AssetSummary
    return { success: true, data: summary }
  } catch (error) {
    console.error('Failed to load asset summary:', error)
    return { success: false, error: 'Failed to load asset summary' }
  }
}

export async function addAsset(input: NewAssetInput): Promise<ApiResponse<number>> {
  try {
    const id = await invoke('add_asset', { input }) as number
    return { success: true, data: id }
  } catch (error) {
    return { success: false, error: typeof error === 'string' ? error : 'Failed to add asset' }
  }
}

export async function updateAsset(
  id: number,
  fields: { name: string; category: string; supplier?: string | null; location?: string | null; note?: string | null },
): Promise<ApiResponse<void>> {
  try {
    await invoke('update_asset', { id, ...fields })
    return { success: true }
  } catch (error) {
    return { success: false, error: typeof error === 'string' ? error : 'Failed to update asset' }
  }
}

export async function disposeAsset(
  id: number,
  proceeds: number,
  method: string | null,
  date: string,
): Promise<ApiResponse<void>> {
  try {
    await invoke('dispose_asset', { id, proceeds, method, date })
    return { success: true }
  } catch (error) {
    return { success: false, error: typeof error === 'string' ? error : 'Failed to dispose asset' }
  }
}

File: src-tauri/src/commands/assets.rs
//! Tauri commands for the fixed-asset register.

use std::sync::Mutex;
use tauri::State;

use crate::db::assets::{self, Asset, AssetSummary, NewAsset};
use crate::state::AppState;

#[tauri::command]
pub fn list_assets(state: State<Mutex<AppState>>) -> Result<Vec<Asset>, String> {
    let mut guard = state.lock().map_err(|_| "state lock poisoned".to_string())?;
    let conn = guard.db.as_mut().ok_or_else(|| "DB not unlocked".to_string())?;
    assets::list_assets(conn).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn get_asset_summary(state: State<Mutex<AppState>>) -> Result<AssetSummary, String> {
    let mut guard = state.lock().map_err(|_| "state lock poisoned".to_string())?;
    let conn = guard.db.as_mut().ok_or_else(|| "DB not unlocked".to_string())?;
    assets::get_asset_summary(conn).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn add_asset(state: State<Mutex<AppState>>, input: NewAsset) -> Result<i64, String> {
    let mut guard = state.lock().map_err(|_| "state lock poisoned".to_string())?;
    let conn = guard.db.as_mut().ok_or_else(|| "DB not unlocked".to_string())?;
    let id = assets::add_asset(conn, &input).map_err(|e| e.to_string())?;
    crate::db::audit::log_audit(
        conn,
        "ASSET_ADDED",
        "asset",
        Some(id),
        &format!("{} — ₹{} ({})", input.name, input.cost, input.funding_method),
    );
    Ok(id)
}

#[tauri::command]
pub fn update_asset(
    state: State<Mutex<AppState>>,
    id: i64,
    name: String,
    category: String,
    supplier: Option<String>,
    location: Option<String>,
    note: Option<String>,
) -> Result<(), String> {
    let mut guard = state.lock().map_err(|_| "state lock poisoned".to_string())?;
    let conn = guard.db.as_mut().ok_or_else(|| "DB not unlocked".to_string())?;
    assets::update_asset(
        conn,
        id,
        &name,
        &category,
        supplier.as_deref(),
        location.as_deref(),
        note.as_deref(),
    )
    .map_err(|e| e.to_string())?;
    crate::db::audit::log_audit(conn, "ASSET_UPDATED", "asset", Some(id), &name);
    Ok(())
}

#[tauri::command]
pub fn dispose_asset(
    state: State<Mutex<AppState>>,
    id: i64,
    proceeds: f64,
    method: Option<String>,
    date: String,
) -> Result<(), String> {
    let mut guard = state.lock().map_err(|_| "state lock poisoned".to_string())?;
    let conn = guard.db.as_mut().ok_or_else(|| "DB not unlocked".to_string())?;
    assets::dispose_asset(conn, id, proceeds, method.as_deref(), &date).map_err(|e| e.to_string())?;
    crate::db::audit::log_audit(
        conn,
        "ASSET_DISPOSED",
        "asset",
        Some(id),
        &format!("proceeds ₹{proceeds}"),
    );
    Ok(())
}
